src/theory/quantifiers/quantifiers_macros.h

```diff
@@ -0,0 +1,78 @@
+/******************************************************************************
+ * Top contributors (to current version):
+ *   Yoni Zohar, Andrew Reynolds, Mathias Preiner
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2021 by the authors listed in the file AUTHORS
+ * in the top-level source directory and their institutional affiliations.
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * Utility for detecting quantifier macro definitions.
+ */
+
+#include "cvc5_private.h"
+
+#ifndef CVC5__THEORY__QUANTIFIERS__QUANTIFIERS_MACROS_H
+#define CVC5__THEORY__QUANTIFIERS__QUANTIFIERS_MACROS_H
+
+#include <map>
+#include <vector>
+#include "expr/node.h"
+
+namespace cvc5 {
+namespace theory {
+namespace quantifiers {
+
+class QuantifiersRegistry;
+
+/**
+ * A utility for inferring macros from quantified formulas. This can be seen as
+ * a method for putting quantified formulas in solved form, e.g.
+ *   forall x. P(x) ---> P = (lambda x. true)
+ */
+class QuantifiersMacros
+{
+ public:
+  QuantifiersMacros(QuantifiersRegistry& qr);
+  ~QuantifiersMacros() {}
+  /**
+   * Called on quantified formulas lit of the form
+   *   forall x1 ... xn. n = ndef
+   * where n is of the form U(x1...xn). Returns an equality of the form
+   *   U = lambda x1 ... xn. ndef
+   * if this is a legal macro definition for U, and the null node otherwise.
+   *
+   * @param lit The body of the quantified formula
+   * @param reqGround Whether we require the macro definition to be ground,
+   * i.e. does not contain quantified formulas as subterms.
+   * @return If a macro can be inferred, an equality of the form
+   * (op = lambda x1 ... xn. def)), or the null node otherwise.
+   */
+  Node solve(Node lit, bool reqGround = false);
+
+ private:
+  bool isBoundVarApplyUf(Node n);
+  bool containsBadOp(Node n, Node op, bool reqGround);
+  bool isGroundUfTerm(Node f, Node n);
+  void getMacroCandidates(Node n,
+                          std::vector<Node>& candidates,
+                          std::map<Node, bool>& visited);
+  Node solveInEquality(Node n, Node lit);
```

Comment:
Can you add documentation? If it was undocumented already in the preprocessing pass then this is a code move and then i realize my request is unfair, but might still be useful :-)
For example, I am not sure what is a "bad op", and what does it mean to be `BoundVarApplyUf`.